File: 2022/day_09.nim
include aoc
import std/sets

type vec2 = array[2, int]

let commands = block:
  var commands: seq[(vec2, int)]

  proc toDelta(dir: char): vec2 =
    case dir:
    of 'U': [0, 1]
    of 'D': [0, -1]
    of 'L': [-1, 0]
    else: [1, 0]

  for line in fetch(2022, 9).splitlines():
    let (_, dir, n) = line.scanTuple("$c $i")
    commands.add (dir.toDelta, n)

  commands

using a, b: vec2

proc `+=`(a: var vec2, b) =
  a[0] += b[0]
  a[1] += b[1]

proc `-`(a, b): vec2 = [a[0] - b[0], a[1] - b[1]]

proc sign(n: int): int =
  if n > 0: 1 elif n < 0: -1 else: 0

proc sign(a): vec2 = [a[0].sign, a[1].sign]

proc chebyshev(a): int = max(a[0].abs, a[1].abs)

proc simulate_rope(nknots: int): int =
  var
    rope = collect(for _ in 0..<nknots: [0, 0])
    seen = toHashSet [[0, 0]]

  for (dir, n) in commands:
    for _ in 0..<n:
      rope[0] += dir
      for i in 0..<nknots - 1:
        let delta = rope[i] - rope[i + 1]
        if delta.chebyshev > 1: rope[i + 1] += delta.sign
      seen.incl rope[^1]

  seen.len

part 1: simulate_rope 2
part 2: simulate_rope 10

File: 2022/day_09.py
import aoc_lube
import numpy as np

def parse_input():
    dirs = dict(D=(1, 0), U=(-1, 0), L=(0, -1), R=(0, 1))
    for line in aoc_lube.fetch(year=2022, day=9).splitlines():
        d, n = line.split()
        yield dirs[d], int(n)

def simulate_rope(nknots):
    rope = np.zeros((nknots, 2), int)
    seen = {(0, 0)}
    for d, n in parse_input():
        for _ in range(n):
            rope[0] += d
            for i in range(nknots - 1):
                delta = rope[i] - rope[i + 1]
                rope[i + 1] += np.sign(delta) * (np.linalg.norm(delta, ord=np.inf) > 1)
            seen.add(tuple(rope[-1]))
    return len(seen)

def part_one():
    return simulate_rope(2)

def part_two():
    return simulate_rope(10)

aoc_lube.submit(year=2022, day=9, part=1, solution=part_one)
aoc_lube.submit(year=2022, day=9, part=2, solution=part_two)
